libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -147,340 +123,606 @@ class BumpPointerAllocator {
+    FunctionDecl *FDSpecialization = FunctionDecl::Create(
+        *ContextAST, ContextAST->getTranslationUnitDecl(), SourceLocation{},
+        DeclarationNameInfo(), VoidFuncType,
+        ContextAST->getTrivialTypeSourceInfo(ContextAST->VoidTy), SC_None,
+        false, false, false, ConstexprSpecKind::Unspecified, nullptr);
+    FDSpecialization->setImplicitlyInline(false);
+
+    FDSpecialization->setDeclName(&ContextAST->Idents.get(KernelName));
+
+    // Used to populate all the template type param decls into.
+    NamespaceDecl *DummyNamespace = NamespaceDecl::Create(
+        *ContextAST, ContextAST->getTranslationUnitDecl(), false,
+        SourceLocation(), SourceLocation(),
+        &ContextAST->Idents.get("DummyNamespace", tok::TokenKind::identifier),
+        nullptr);
+    DummyNamespace->setImplicit(true);
+    ContextAST->getTranslationUnitDecl()->addDecl(DummyNamespace);
+
+    // Will be used to build template parameter list.
+    SmallVector<NamedDecl *> TemplateNamedDecls;
+    // Used for setting template specialisation.
+    SmallVector<TemplateArgument> TemplateArguments;
+    // Used for the fixups.
+    SmallVector<clang::QualType> TemplateTypeParamTys;
+    unsigned TemplateIndex = 0;
+    for (auto &TemplateArgQT : TemplateArgTys) {
+      std::string const Name{std::string{"TempTy"} +
+                             std::to_string(TemplateIndex)};
+      auto &II = ContextAST->Idents.get(Name);
+      auto *TTPD = TemplateTypeParmDecl::Create(
+          *ContextAST, DummyNamespace, SourceLocation(), SourceLocation(), 0,
+          TemplateIndex, &II, /* Typenaem */ true, /*ParameterPack*/ false);
+      TTPD->setDefaultArgument(
+          ContextAST->getTrivialTypeSourceInfo(TemplateArgQT));
+
+      TemplateNamedDecls.emplace_back(TTPD);
+      auto TA = TemplateArgument(TemplateArgQT);
+      TemplateArguments.emplace_back(TA);
+
+      // Store this qualified type with newly created proper template type
+      // param qualified type.
+      TemplateTypeParamTys.push_back(
+          ContextAST->getTemplateTypeParmType(0, TemplateIndex, false, TTPD));
+
+      ++TemplateIndex;
+    }
+    // Fix up the template types in the original FD's arg tys and return ty.
+    auto AreQTsEqual = [&](clang::QualType &LHS, clang::QualType &RHS) -> bool {
+      return (LHS.getBaseTypeIdentifier() && RHS.getBaseTypeIdentifier() &&
+              LHS.getBaseTypeIdentifier()->isStr(
+                  RHS.getBaseTypeIdentifier()->getName())) ||
+             LHS == RHS;
+    };
+    unsigned NumReplaced = 0;
+    unsigned Idx = 0;
+    for (auto &TemplateArgQT : TemplateArgTys) {
+      if (AreQTsEqual(TemplateArgQT, RetTy)) {
+        RetTy = TemplateTypeParamTys[Idx];
+        goto Found;
+      }
+      for (unsigned i = 0; i < ArgTys.size(); ++i) {
+        if (AreQTsEqual(ArgTys[i], TemplateArgQT)) {
+          ArgTys[i] = TemplateTypeParamTys[Idx];
+          goto Found;
+        }
+      }

```

Comment:
Are you basically trying to instantiate the function here?

Reply:
Yeap, making sure that the template is fully specialized.

Reply:
Ooofda, I think you might need to use the usual template instantiation machinery to do this (you may have multiple levels of instantiation to perform for a type like `foo<T<U>>` (think: template template parameters). CC @erichkeane

Reply:
This is a bit of a tricky issue.
In principle you are right; the following, fictional, overload of image read would not be handled correctly:

```cpp
template <typename T> class woof{};
template <>
int4 __spirv_ImageRead<int4 , ocl_image3d_ro, woof<int4>>(ocl_image3d_ro, woof<int4>) {}
```

However, this patch:
* does not deviate from the current status, the same failure could be observed in the mainline,
* we only need to be able to handle functions defined in libclc, all of which have to follow spir-v interface,
* I am concerned that if we wanted to make it complete, we would end up re-implementing huge chunks of clang for something that is a closed scope task,
* finally, remangler will hard fail should the interface changed.

With thte patch, the above sample would result in: 

```sh
Unhandled type : woof<int vector[4]>
UNREACHABLE executed at /home/jakub/Projects/oneAPI/jchlanda/llvm/src/llvm/frontendaction/LibclcRemangler.cpp:479!
```

Reply:
Ooof, there are quite a lot of situations that'll get out of hand here, its not handling constraint overloading, template-template params, etc, but I don't have a good hold on it.  BUT if the libclc owner is equally unconcerned as you about these, I've got no problems obviously.

Reply:
@steffenlarsen would you be able to weigh in on this?

Reply:
Use of templates in the current builtins are luckily limited and hopefully they stay this way. It would be nice to see handling of these improve in the future (hopefully in a way where it can again reuse some existing logic) but at least this doesn't seem to worsen the handling of these outliers compared to mainline.

Note that the builtins libclc attempts to mimic need to be handled correctly by the SPIR-V Translator too, so to my knowledge the subset of possible instantiations we support of the templated builtins are actually fairly limited too.